[CUL-85] 카카오 소셜 로그인 구현

euncheol-kim opened the pull request:
## 이슈 번호, 혹은 추가한 기능에 대해 설명해주세요.

[[BE] 사용자가 카카오 소셜 로그인을 할 수 있도록 한다. #CUL-85](https://chopinoff.atlassian.net/browse/CUL-85?atlOrigin=eyJpIjoiOWNkOGRmYzdhOTYwNDkyNWFlNTlhNDJlMWI1ZmJjMzIiLCJwIjoiaiJ9)

## 작업 상세 내용

- [ ] 사용자가 카카오 로그인을 할 수 있도록 한다.

## 기타 내용
[CultureSpot-Backend-Secrets](https://github.com/culturespot/CultureSpot-Backend-Secrets)의 아래 내용이 변경되었습니다.
- 변경 내용
  - `CultureSpot-ServiceApi/src/main/resources/application.yml`
    - Kakao 소셜 로그인 도입으로 인해 설정과 관련한 변경 발생
  - `env/serviceApi-module.env`
    - Kakao 소셜 로그인 도입으로 클라이언트 id의 추가
  - `env/domain-module.env`
    - `GOOGLE_CLIENT_ID`, `GOOGLE_CLIENT_SECRET` 위치가 부적합하여 삭제 후 `env/serviceApi-module.env`로 이동 (업데이트 하기 전부터 중복으로 존재했습니다) 
 - 변경 사유
   -  Kakao 클라이언트 id 추가로 인한 변경 발생

euncheol-kim committed:
[comment💡] Kakao 로그인 주석 추가

euncheol-kim committed:
[feat✨] AuthExceptionCode 추가

euncheol-kim committed:
[refactor♻️] 소셜 로그인 principal을 email로 사용하도록 조치(#CUL-175)

# 카카오 로그인 도입전에는 문제가 없었지만 카카오 로그인을 도입하고 문제가 발생하여 코드 변경 조치

euncheol-kim committed:
[remove🔥] EndpointType 필요없는 코드 삭제

# API 할당 방식이 정해지면 삭제하거나 유지할 예정

euncheol-kim committed:
[feat✨] ExceptionStatusMapper 수정

# ExceptionStatusMapper는 GlobalException에서 사용합니다.

euncheol-kim committed:
[feat✨] 사용자 소셜 로그인 취소에 대한 응답 리스폰스 추가

# 해당 코드가 없으면 SecurityFilter에 의해 403으로 반환되어 추가해주었습니다.

euncheol-kim committed:
[feat✨] 사용자 소셜 로그인 성공에 대한 응답 리스폰스 추가

# 프론트 측에서 작성해준 초기 API 명세 요구 조건을 따랐습니다.

euncheol-kim committed:
[refactor♻️] 로그인 성공시 JSON 포맷 형식의 응답을 하도록 변경

# 프론트 측에서 작성해준 API 명세 요구 조건을 따랐습니다.

euncheol-kim committed:
[feat✨] 소셜 로그인 구분에 따른 핸들 메소드 작성

# 프론트 측에서 작성해준 API 명세 요구 조건을 따랐습니다.

euncheol-kim committed:
[feat✨] 소셜로그인 회원가입&로그인 실패시 동작하는 핸들러 추가

euncheol-kim committed:
[refactor♻️] 소셜로그인 회원가입&로그인 실패/취소시 동작하는 핸들러 추가에 따른 SecurityFilter 변경

euncheol-kim committed:
[comment💡] UserService, UserServiceImpl 주석 추가

euncheol-kim committed:
[refactor♻️] User 엔티티 profile code 추가 및 userId필드 -> id로 변경

# id필드 변경에 따른 PostServiceImpl 파일의 수정 발생

euncheol-kim committed:
[chore⚙️] domain 모듈 - application, application-prod 최신화

##
# 프로파일들이 최신화 되어 있지 않아 커밋합니다.
##

euncheol-kim committed:
[refactor♻️] domain 모듈 - 잘못된 병합으로 인한 코드 수정

##
# 해당 커밋 전, 가장 최근에 작업해주신 내역을 반영했습니다.
##

euncheol-kim committed:
[feat✨] 사용자 최신 로그인 시간 필드와 업데이트 기능 추가

euncheol-kim committed:
[refactor♻️] 성공적인 소셜 로그인시, 사용자의 최신 로그인 시간 업데이트 로직 추가

spearkkk commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/entity/User.java:
```diff
@@ -28,9 +30,15 @@ public class User extends BaseEntity {
     private SocialLoginType authType;
+
+    @Column(name="last_login_at", nullable = false)
+    private LocalDateTime lastLoginAt = LocalDateTime.now();
```
이 부분에서 값을 초기화하지 말고 User 객체를 생성할 때, 넣어줄 수 있을까요?

euncheol-kim replied:
아 네 객체를 만들 때 넣어줄 수 있도록 조치하겠습니다.

spearkkk commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/service/UserService.java:
```diff
@@ -1,12 +1,44 @@
+     */
+    User findUserOrThrow(String email);
+
+    /**
```
주석에 대해서 다양한 의견이 있지만, 특별하게 고려해야할 부분이 아니라면 간략하게 유지해도 좋아보여요.
예를 들어, `@param email` 이 없어도 코드 읽을 때 알 수 있는 내용이라 주석으로 인한 장점이 명확해보이지는 않아 보입니다.

euncheol-kim replied:
아 네 알겠습니다. 명확히 알아볼 수 있는 건 주석을 지워 PR에 다시 반영하겠습니다. 감사합니다!

OhHaneol commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/service/UserService.java:
```diff
@@ -1,12 +1,44 @@
 public interface UserService{
+    /**
+     *  사용자가 갖고 있는 권한 조회
+     * <p>
```
저도 마찬가지로 주석 내부에 <p> 태그가 다른 역할을 하지 않는다면 간략하게 유지해도 좋지 않을까 합니다.

ohahsis reviewed (approved):
고생하셨습니다! 저도 마찬가지로 주석을 간략화하면 좋을 것 같습니다. (예를 들어 `<p>`태그)

euncheol-kim commented:
@OhHaneol 

![image](https://github.com/user-attachments/assets/ae662f53-ee90-426a-abaf-7b20ce9d9e26)
제가 작성한 주석중 일부인데요...! 만약 `<p> 권한이름은.... 반환됩니다 </p>`가 없으면 `권한 조회 권한이름은....반환됩니다.`가 붙어버리는 문제가 있습니다! 그래서 p태그의 경우는 필요할 것 같아요 ㅠㅠ..

euncheol-kim committed:
[refactor♻️] CustomOAuth2UserService.java 파일 중복코드 리팩토링 진행

# createNewUser(email, authType)내부 createUser 변수에서 사용자 정보를 저장하는 로직 수정

euncheol-kim committed:
[refactor♻️] OAuth2LoginHandler -> OAuth2AuthenticationSuccessHandler 이름변경

# OAuth2AuthenticationSuccessHandler: 이름 변경에 따른 타입 수정

euncheol-kim committed:
[refactor♻️] User객체의 lastLoginAt필드 초기화 방식 수정 -> 객체 생성시 값을 정하도록 변경

# lastLoginAt = LocalDate.now() 방식에서 Builder 방식으로 객체 생성시 값을 정하도록 변경

euncheol-kim committed:
[refactor♻️] User 객체 생성 코드 분리

# User 객체 생성 코드 분리에 따른 createUserIfNotExists(String email, SocialLoginType authType) 내부 로직 변경

euncheol-kim committed:
[feat✨] 주석 수정 및 User객체 생성 추상 메소드 선언

euncheol-kim commented:
@OhHaneol @spearkkk

[V] User 객체를 생성시 필드 last_login_at 할당할 수 있도록 조치
 - 기존에 LocalDate.now() 초기화 방식 제거
[V] 변수명, 네이밍으로 명확히 유추 가능한 주석은 제거

다시 한 번 검토해주시면 감사하겠습니다..!

ohahsis commented:
> @OhHaneol
> 
> ![image](https://private-user-images.githubusercontent.com/187013193/430237959-ae662f53-ee90-426a-abaf-7b20ce9d9e26.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM4MTIyNjgsIm5iZiI6MTc0MzgxMTk2OCwicGF0aCI6Ii8xODcwMTMxOTMvNDMwMjM3OTU5LWFlNjYyZjUzLWVlOTAtNDI2YS1hYmFmLTdiMjBjZTlkOWUyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwNVQwMDEyNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YWEyNjRkODNmMGJjNjk3ZDZmZjk0NmY5ZDIwZWQ3NDg2ZmQwMWI4NjExMDgwMTA0NDA2NjQ5NDJjZjFjNTYzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.uOo-nTjqxovQIjQ7q8Vb5RrW9q-_OCUtJ5SCfy3-6H0) 제가 작성한 주석중 일부인데요...! 만약 `<p> 권한이름은.... 반환됩니다 </p>`가 없으면 `권한 조회 권한이름은....반환됩니다.`가 붙어버리는 문제가 있습니다! 그래서 p태그의 경우는 필요할 것 같습니다. 혹시 더 나은 방식이 있을까요? 주석도 연습으로 좀 작성한거거든요..!

아하...! 이유 확인했습니다! 이미지를 첨부해주셔서 알기 쉬웠습니다 감사해요☺️
